bazel: update macros to support multi-arch and create a CROSSTOOL for gcc crossbuild-essential packages

ixdy opened the pull request:
This PR is basically 2-3 changes combined, but it probably makes sense to look at them together, since all of them are needed for the bazel cross-compilation PR in kubernetes/kubernetes (https://github.com/kubernetes/kubernetes/pull/73930). (I can separate them if you feel very strongly, however.)

1. Support `select()` in `release_filegroup`: I achieve this by adding a separate attribute (`conditioned_srcs`), creating hash rules for all possible outputs, and then using `select()`s to only reference the correct hashes for the platform that's been configured.
2. Add a `goarch` attribute to the `k8s_deb` and `deb_data` macros and create a new macro for `pkg_rpm` which translate to the corresponding architecture name expected for that package type. For example, `arm64` is in some places instead referred to as `aarch64`, so handle the translation transparently.
3. Create a Go program to generate a `CROSSTOOL` file, and then check this generated `CROSSTOOL` file in. I initially generated this file by running `bazel build` inside a debian container with `gcc` installed. I took the resulting auto-generated `CROSSTOOL`, copied it, removed parts that didn't seem relevant, and figured out which pieces needed to be overridden with the gcc-cross binaries and libraries.

The `CROSSTOOL` bit is somewhat fragile, since it basically requires the `crossbuild-essential-PLATFORM` packages to be installed. It also can't be auto-generated at build time, since I believe it's used in the analysis phase, and I didn't want to create a `repository_rule` (at least not for this first pass).

It is possible to use `clang` instead, and in that case it'd even be possible to automatically download `clang` as a dependency, but we still need a sysroot (libc, includes, etc) for each platform, and those don't seem to be obviously available anywhere. (The chromium project has some prebuilt sysroots, but not one for every platform we need.)

An example of using this `CROSSTOOL` on my linux machine:

```
bazel build --crosstool_top=@io_k8s_repo_infra//tools:toolchain --cpu=arm --compiler=gcc  --platforms=@io_bazel_rules_go//go/toolchain:linux_arm //cmd/kubectl
```

/assign @BenTheElder @fejta 
cc @liztio

fejta commented on defs/build.bzl:
```diff
@@ -71,7 +71,7 @@ gcs_upload = rule(
 
 # Computes the md5sum of the provided src file, saving it in a file named 'name'.
-def md5sum(name, src, visibility = None):
+def md5sum(name, src, tags = None, visibility = None):
```
Use **kwargs for tags, visibility, etc?

ixdy replied:
done

fejta commented on defs/build.bzl:
```diff
@@ -134,7 +137,7 @@ def _hashes_for_srcs(srcs, srcs_basenames_needing_hashes):
 # conditioned_srcs is a dictionary mapping conditions to label lists.
 #   It will be passed to select().
-def release_filegroup(name, srcs = None, conditioned_srcs = None, visibility = None):
+def release_filegroup(name, srcs = None, conditioned_srcs = None, tags = None, visibility = None):
```
Is this part of another commit? I don't see where k8s_deb or pkg_rpm call this (and/or now use tags)?

ixdy replied:
release_filegroup isn't used directly in any rules here, but it's used in a number of places over in k/k.

fejta commented on tools/CROSSTOOL:
```diff
@@ -0,0 +1,510 @@
+# limitations under the License.
+
+# DO NOT EDIT
+# This file was generated by generate_crosstool
```
Can we make it note the proto message type it is outputting?

ixdy replied:
done.

fejta commented on WORKSPACE:
```diff
@@ -13,3 +13,9 @@ load("@io_bazel_rules_go//go:deps.bzl", "go_register_toolchains", "go_rules_depe
+http_archive(
+    name = "io_bazel",
+    sha256 = "6860a226c8123770b122189636fb0c156c6e5c9027b5b245ac3b2315b7b55641",
+    url = "https://github.com/bazelbuild/bazel/releases/download/0.22.0/bazel-0.22.0-dist.zip",
```
Interesting... will this work with future versions of bazel?

ixdy replied:
The proto definition is changing (see https://github.com/bazelbuild/bazel/issues/6861), and we'll likely have to migrate at some point. It might even be possible to migrate right now and have things work back to at least Bazel 0.21.0, but I wanted to start somewhere and then iterate.

Also, eventually this is supposed to be all Starlark, which should be much easier to generate.

fejta commented on tools/BUILD.bazel:
```diff
@@ -0,0 +1,60 @@
+    name = "gen_crosstool",
+    # This is a proto, not a bzl file, but close enough.
+    srcs = ["//verify/boilerplate:boilerplate.bzl.txt"],
+    outs = ["CROSSTOOL"],
```
`CROSSTOOL.generated.pb.txt`?

ixdy replied:
I went with `generated_CROSSTOOL.textpb`.

fejta commented on tools/BUILD.bazel:
```diff
@@ -0,0 +1,60 @@
+# Note: this isn't actually generated at build time, but you can run this rule
+# and then copy the generated CROSSTOOL file into the tree.
```
Plesae make hack/update-crosstool.sh and hack/verify-crosstools.sh scripts to do this

Maybe make make it bounce if bazel is not on the machine unless a flag is set (which set set for ci). 

Alternatively how about a sh_binary (to copy over the generated file to tree) and sh_test (to check that the generated file is the same as what is in tree).

That way the sh_test will only run in the bazel test //... scenario.

ixdy replied:
I like the `sh_test` approach - it can be as easy as `diff generated_file checked-in_file`. Will work on that.

ixdy commented:
The other thing to note is that some `CROSSTOOL` fields are being removed (https://github.com/bazelbuild/bazel/issues/6861) and eventually being replaced by starlark. I haven't processed this file through the migrator tool, since it's not clear whether the new fields work with bazel older than 0.22.0. We might want to fix that at some point, though.

fejta commented on tools/BUILD.bazel:
```diff
@@ -0,0 +1,60 @@
+# Note: this isn't actually generated at build time, but you can run this rule
```
Can you comment why this isn't generated at build time? Limitation or intentional decision (why)?

ixdy replied:
done

fejta commented on tools/generate_crosstool/main.go:
```diff
@@ -0,0 +1,229 @@
+)
+
+var (
+	out         = flag.String("out", "", "filename for CROSSTOOL text proto to write")
```
Writing the proto to stdout seems easier and more flexible here

ixdy replied:
you can always do `-out=/dev/stdout` if you want it on stdout. 

it felt more natural to write out to a file directly rather than rely on shell redirection.

BenTheElder commented on defs/deb.bzl:
```diff
@@ -19,29 +19,39 @@ KUBERNETES_AUTHORS = "Kubernetes Authors <kubernetes-dev+release@googlegroups.co
 
-def k8s_deb(name, **kwargs):
+GOARCH_TO_DEBARCH = {
+    "386": "i386",
```
do we build kubernetes for i386??

ixdy replied:
yes. https://github.com/kubernetes/kubernetes/blob/805a9e703698d0a8a86f405f861f9e3fd91b29c6/hack/lib/golang.sh#L137-L148

(in reality, we probably only care about windows/386. https://github.com/kubernetes/kubernetes/issues/28632)

mikedanese commented on BUILD.bazel:
```diff
@@ -1,5 +1,8 @@
 # gazelle:prefix k8s.io/repo-infra
 
+# Use the Bazel-vendored protobuf library since we use go_proto_library
+# gazelle:resolve go github.com/golang/protobuf/proto @com_github_golang_protobuf//proto:go_default_library
```
Do you want this?
```
# gazelle:proto disable_global
```

ixdy replied:
Yes, because I'm using `go_proto_library` in `tools/generate_crosstool/BUILD.bazel` to build the go library for bazel's crosstool proto.

mikedanese commented on defs/build.bzl:
```diff
@@ -123,25 +121,72 @@ def release_filegroup(name, srcs, visibility = None):
+        sha512sum(name = basename + ".sha512", src = src, tags = tags, visibility = visibility)
+
+    if conditioned_srcs:
+        native.filegroup(
```
Should we be using output_group instead?

https://docs.bazel.build/versions/master/be/general.html#filegroup.output_group

ixdy replied:
I'm not sure how that helps us?

ixdy commented:
Travis is failing because this `generate_crosstool` requires bazel and the go linters don't know how to handle that. hm...

ixdy commented:
finally got travis to pass by adding a `go_path` rule and using that with `go vet`. 

I think this is ready for another review. PTAL?

I'll squash back down after review.

fejta commented on verify/BUILD.bazel:
```diff
@@ -0,0 +1,20 @@
+    name = "verify-crosstool",
+    srcs = ["verify-crosstool.sh"],
+    args = [
+        "$(location //tools:CROSSTOOL)",
```
Can we make CROSSTOOL also have a .textpb extension?

ixdy replied:
nope, it must be named `CROSSTOOL`.

fejta commented on verify/verify-crosstool.sh:
```diff
@@ -0,0 +1,30 @@
+diff=$(diff -u "${checked_in_crosstool}" "${generated_crosstool}" || true)
+
+if [[ -n "${diff}" ]]; then
+  echo "Checked-in CROSSTOOL does not match generated CROSSTOOL"
```
Can we get a command here I need to run to update it?

ixdy replied:
done

ixdy commented:
ok, squashed back to 3 commits. I've also removed the gcc version from the include paths, so we should be compatible with more versions of debian (stretch, buster, and whatever google workstations are using).

fejta commented:
/lgtm

ixdy commented:
/hold cancel
